src/include/miopen/solver/implicitgemm_ck_util.hpp

```diff
@@ -0,0 +1,115 @@
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ *
+ *******************************************************************************/
+
+#pragma once
+
+#include <miopen/conv/data_invoke_params.hpp>
+
+namespace miopen {
+namespace solver {
+
+template <typename ConvPtrsType>
+typename ConvPtrsType::iterator FindConvPtrByID(ConvPtrsType& conv_ptrs,
+                                                const std::string& kernel_id)
+{
+    return std::find_if(conv_ptrs.begin(), conv_ptrs.end(), [&kernel_id](const auto& ptr) {
+        return ptr->GetTypeString() == kernel_id;
+    });
+}
+
+template <typename DeviceOpType, typename CKArgsType>
+std::vector<std::string> FillValidKernelsIDs(const ProblemDescription& problem)
+{
+    const auto args      = CKArgsType{problem};
+    const auto conv_ptrs = DeviceOpType::GetInstances();
+    assert(!conv_ptrs.empty());
+
+    std::vector<std::string> valid_kernels;
+    valid_kernels.reserve(conv_ptrs.size());
+    for(size_t idx = 0; idx < conv_ptrs.size(); ++idx)
+    {
+        if(args.IsSupportedBy(conv_ptrs[idx]))
+            valid_kernels.emplace_back(std::move(conv_ptrs[idx]->GetTypeString()));
+    }
+    assert(!valid_kernels.empty());
+    return valid_kernels;
+}
+
+template <typename DeviceOpType, typename CKArgsType>
+bool IsCKArgsSupported(const ProblemDescription& problem, const std::string& kernel_id)
+{
+    auto conv_ptrs = DeviceOpType::GetInstances();
+    auto ptr_iter  = FindConvPtrByID(conv_ptrs, kernel_id);
+
+    return (ptr_iter != conv_ptrs.end()) && CKArgsType{problem}.IsSupportedBy(*ptr_iter);
+}
+
+template <typename DeviceOpType, typename CKArgsType>
+bool IsCKApplicable(const ProblemDescription& problem)
+{
+    const auto args = CKArgsType{problem};
+    if(!std::all_of(args.strides.begin(), args.strides.end(), [](auto x) { return x == 1; }))
+        return false;
+
+    const auto ptrs = DeviceOpType::GetInstances();
+    return std::any_of(
+        ptrs.begin(), ptrs.end(), [&args](auto& ptr) { return args.IsSupportedBy(ptr); });
+}
+
+template <typename DeviceOpType, typename CKArgsType>
+ConvSolution InitInvokerFactory(const ProblemDescription& problem, const std::string& kernel_id)
+{
+    auto conv_ptrs = DeviceOpType::GetInstances();
+    auto ptr_iter  = FindConvPtrByID(conv_ptrs, kernel_id);
+
+    if(ptr_iter == conv_ptrs.end())
+        MIOPEN_THROW("PerformanceConfig kernel '" + kernel_id + "' does not exist");
+
+    ConvSolution result;
+    result.invoker_factory =
+        [ck_args     = CKArgsType{problem},
+         sh_conv_ptr = std::shared_ptr{std::move(*ptr_iter)}](const std::vector<Kernel>&) mutable {
+            return [ck_args = std::move(ck_args), sh_conv_ptr = std::move(sh_conv_ptr)](
+                       const Handle& handle, const AnyInvokeParams& primitive_parameters) {
+                const auto& data_ctx = primitive_parameters.CastTo<conv::DataInvokeParams>();
```

Comment:
This DataInvokeParams only works for fwd and backward data. For backward weight, we need WrWInvokeParams. I would suggest to either template it or check the convolution direction to decide which Invokeparams to be called. Sth like:
if(problem.direction.IsBackwardWrW())
   data_ctx = primitive_parameters.CastTo<conv::WrWInvokeParams>();
else
   data_ctx = primitive_parameters.CastTo<conv::DataInvokeParams>();